Frontend Migration mittel der Chicken-Little-Methode (#135)
adessoAG/devblog: _posts/2020-03-16-frontend-migration-mittels-der-chicken-little-methode.md

## Technischer Hintergrund
Das Versicherungsportal basiert maßgeblich auf den Technologien *Java EE 6* sowie dem Webframework [*Seam*](http://www.seamframework.org/) in der Version 2.
Bereit gestellt wird die Anwendung auf einem **JBoss Application Server** in der *Version 5*.

Die Entwicklung von *Seam* wurde bereits im Jahr 2012 vollständig eingestellt.
Dieser Umstand führt unter anderem dazu, dass in der Weiterentwicklung des Projekts der Einsatz neuerer Technologien dadurch limitiert wird, dass die Kompatibilität zu *Seam* immer gewährleistet sein muss.
Davon betroffen sind sowohl die eingesetzte Java-Version, damit einhergehend auch der Applikationsserver, als auch schlussendlich die verwendeten *Libraries*.
Damit qualifiziert sich dieses Projekt, um im Kontext der **Softwaremigration** näher betrachtet zu werden.

Das in diesem Artikel betrachtete Beispiel entstammt einer Abschlussarbeit, die durch die *adesso SE* betreut wurde.  
Daher wurde auch die Entscheidung getroffen, eine *Frontend-Migration* durchzuführen, um die Komplexität überschaubar zu halten und gleichzeitig den Rahmen der erwähnten Arbeit nicht zu sprengen.

> Schlussendlich demonstriert dieses Beispiel aber auch das breite Einsatzgebiet dieser Methode, die sich flexibel für entweder nur eine Migrationsart (Programm-, Benutzerschnittstellen- und/oder Datenmigration), eine Kombination daraus oder aber auch für vollständige Systemmigrationen, die sämtliche Bereiche umfassen, einsetzen lässt.

Reviewer: Der Satz ist sehr komplex aufgebaut. Vielleicht kann der Satz so in etwa aufgesplittet werden? 
```suggestion
Schlussendlich demonstriert dieses Beispiel aber auch das breite Einsatzgebiet dieser Methode. Sie lässt sich flexibel für entweder nur eine Migrationsart (Programm-, Benutzerschnittstellen- und/oder Datenmigration) oder eine Kombination dieser einsetzen. Eine vollständige Systemmigration, die sämtliche Bereiche umfasst, ist aber auch möglich.
```

# Die Migration

In diesem Abschnitt erfolgt die Migration der zuvor beschriebenen Anwendung.
Da hier zunächst nur das Frontend migriert werden soll, entfallen von den ursprünglich [elf Schritten](/_posts/2020-03-16-inkrementelle-softwaremigration-nach-der-chicken-little-methode.md#die-chicken-little-methode) der **Chicken-Little-Methode**, all jene Punkte, die sich auf die Bereiche **Programme und Anwendungen** sowie die **Datenhaltung** beziehen.

> Übrig bleiben also lediglich noch die **Analyse** _(Schritt 1)_ sowie die **Strukturierung** _(Schritt 2)_ des Altsystems, der **Entwurf** und die **Installation der Zielumgebung** _(Schritt 3 + 6)_ und die **Installation der benötigten Schnittstellen** _(Schritt 7)_ sowie die abschließende **Migration und Umstellung** auf die Neuentwicklung _(Schritt 10 + 11)_.

Reviewer: Hier werden 5 inhaltsreiche Punkte in einem Satz aufgezählt. Weiter unten wird bei ähnlichem inhaltlichem Umfang bei drei Punkten bereits eine Aufzählungsliste verwendet. Hier könnte eine solche Liste der übersicht halber auch gut verwendet werden.

            <h:outputText value="#{_vertrag.versicherer()}" />
        </rich:column>
        ...
    </rich:dataTable>
    ...
    <h:commandButton id="berechnenButton"
        action="#{vertragslisteController.berechneVertraege()}"
        value="Berechnen" />

</ui:define>
```

#### Involvierte Java-Klassen
Aus dem oben stehenden *JSF-Template* lassen sich bereits die beteiligten *Java-Klassen* ablesen.
Die zentralen Elemente sind hierbei der Controller `VertragslisteController`, der das Datenhandling übernimmt, sowie die Entität `Vertrag`, die den jeweiligen Versicherungsvertrag repräsentiert und dessen Details bereitstellt.

## Entwurf und Entwicklung der neuen Benutzeroberfläche

Die Darstellung der *Vertragsliste* innerhalb der *React App* soll auch hier wieder in tabellarischer Form erfolgen.  

Sowohl die *Vertragsliste* als auch der *Vertrag* werden dazu jeweils als eigenständige *React Komponente* innerhalb der *React App* implementiert.
Bei einer *React Komponente* handelt es sich knapp zusammengefasst um einen wiederverwendbaren Teil einer Benutzeroberfläche, der die Funktionalität sowie die Eigenschaften der Komponente (`props`) innerhalb einer *React App* bündelt.

> In ihrem Aufbau ähnelt sie stark dem von *(Java-)Klassen*.  

Reviewer: ```suggestion
In ihrem Aufbau ähnelt sie *(Java-)Klassen* stark.  
```
oder
```suggestion
Ihr Aufbau ähnelt stark dem einer *(Java-)Klasse*.  
```

    componentDidMount() {}

    // Vertrag im Altsystem löschen
    remove(vertragsId) {}

    // Berechnung der Tarife der i.d. Liste enthaltenen Verträge anstoßen
    calculate() {}

    // Verarbeitung der einzelnen Verträge ...
    eachVertrag(vertrag, i) {}
     // ... und anschließende Ausgabe der Vertragsliste im Frontend
         render() {}
}
```

Im *Konstruktor* erfolgt zunächst die Initialisierung des *Zustands* ([`state`](https://reactjs.org/docs/state-and-lifecycle.html)) der Vertragslisten-Komponente.

Von wesentlicher Bedeutung ist hier das zunächst leere Array `vertrage`, in welchem die, über die Schnittstelle zum Altsystem abgefragten Versicherungsverträge für die jeweilige *View* gespeichert werden.
Daneben gibt es noch zwei weitere Felder, die kennzeichnen, ob eine Anfrage an die Schnittstelle noch läuft oder erfolgreich abgeschlossen wurde (`requestComplete`).
Scheitert der *Request*, so wird der aufgetretene Fehler in `error` gespeichert, um diesen im weiteren Verlauf behandeln zu können.   
Des Weiteren erfolgt hier das [*Binding*](https://reactjs.org/docs/faq-functions.html#why-is-binding-necessary-at-all) der Methoden an den Wert `this`.  

> Die initiale Abfrage der *Vertragsliste* im Altsystem erfolgt später über die noch zu implementierende Schnittstelle mittels der Methode [`componentDidMount()`](https://reactjs.org/docs/react-component.html#componentdidmount).

Reviewer: ```suggestion
Die initiale Abfrage der *Vertragsliste* au dem Altsystem erfolgt später über eine noch zu implementierende Schnittstelle mittels der Methode [`componentDidMount()`](https://reactjs.org/docs/react-component.html#componentdidmount).
```

 Hierbei handelt es sich um eine sogenannte [*Lifecycle-Methode*](https://reactjs.org/docs/state-and-lifecycle.html#adding-lifecycle-methods-to-a-class) von *React*, die ausgeführt wird, sobald die Komponente der Benutzeroberfläche hinzugefügt wurde.

> Weitere Methoden sind hier `remove()` für das Löschen von Verträgen, sowie `calculate()` zum Anstoßen der Berechnung sämtlicher Tarife.

Reviewer: ```suggestion
Weitere Methoden in dieser Komponente sind hier `remove()` für das Löschen von Verträgen, sowie `calculate()` zum Anstoßen der Berechnung sämtlicher Tarife.
```

                           this.remove()
                       }>Löschen</a>
                </td>
            </tr>
        );
    }
}

export default Vertrag;
```      

Die Ausgabe der jeweiligen Tabellenzeile erfolgt in der `render()`-Methode.
Die jeweiligen Vertragsdetails wurden bei der Initialisierung der Komponente `Vertrag` innerhalb der `Vertragsliste`, in Form von `JSX` übergeben.
Auf diese kann nun über das Objekt `this.props.children` zugegriffen werden.
Daneben erfolgt hier außerdem noch die Definition der Buttons, die für das *Löschen* sowie das *Bearbeiten* des jeweiligen Vertrages verantwortlich sind.

Mit der Entwicklung der beiden Komponenten für die *Vertragsliste* und den *Vertrag* ist die neue Benutzeroberfläche fertiggestellt.

Im nächsten Schritt müssen nun noch die **Schnittstellen** im Altsystem implementiert werden, damit die hier entwickelte *React App* und das Altsystem miteinander kommunizieren können.

## Entwicklung einer REST-API als Schnittstelle zwischen Alt und Neu
Die Implementierung der Schnittstelle erfolgt in diesem Beispiel in Form einer kleinen [**REST-API**](https://de.wikipedia.org/wiki/Representational_State_Transfer). 

> Die **API** soll hierbei als **Reverse Gateway** dienen, der die eingehenden **Requests** an die original Methode des **Vertragslisten Controller** delegiert und mögliche Rückgaben wieder an den Aufrufer weiterleitet.   

Reviewer: ```suggestion
Die **API** soll hierbei als **Reverse Gateway** dienen, der die eingehenden **Requests** an die ursprüngliche Methode des **Vertragslisten Controller** delegiert und mögliche Rückgaben wieder an den Aufrufer weiterleitet.   
```
oder 
```suggestion
Die **API** soll hierbei als **Reverse Gateway** dienen, der die eingehenden **Requests** an die Originalmethode des **Vertragslisten Controller** delegiert und mögliche Rückgaben wieder an den Aufrufer weiterleitet.   
```

    }
}
```      

Für die Umsetzung wird sich der Library **RESTEasy** bedient, die den **JAX-RS**-Standard implementiert und die Definition und Konfiguration von *REST-Ressourcen* über Annotationen der *Java-Klassen* erlaubt. 

Über dem Klassennamen erfolgt zunächst die Definition der URI `/vertragsliste` per Annotation.
Darüber lässt sich die Ressource in der **REST-API** adressieren.
Daneben wird außerdem festgelegt, dass die Kommunikation via `JSON` erfolgen soll. 

Die folgende Tabelle listet die beteiligten **Ressourcen**, die jeweilige **URI**, die dabei verwendete **HTTP-Methode** sowie die jeweils korrespondierende **Java-Methode** auf.

| Aktion                 | HTTP-Methode       | URI                           | Java-Methode           |
| ---------------------- | ------------------ | ----------------------------- | ---------------------- | 
| **Liste abfragen**     | GET                | `/vertragsliste`              | `getVertragsliste()`   |
| **Verträge berechnen** | POST               | `/vertragsliste`              | `berechneAngebote()`   |
| **Vertrag löschen**    | DELETE             | `/vertragsliste/{vertragsId}` | `removeVetrag()`       |

> Für sowohl das **Abfragen** der Liste als auch die **Berechnung** der Verträge zeichnet sich die Ressource `Vertragsliste` verantwortlich.  

Reviewer: ```suggestion
Sowohl für das **Abfragen** der Liste als auch die **Berechnung** der Verträge zeichnet sich die Ressource `Vertragsliste` verantwortlich.  
```

</ui:define>
```

Mit diesem letzten Schritt ist die Migration der Vertragsliste abgeschlossen und kann im Altsystem nun verwendet werden.

# Zusammenfassung

Dieser Artikel hat gezeigt, wie sich die **Chicken-Little-Methode** für eine Frontend-Migration anwenden lässt. Dazu wurde das Beispiel der **Vertragsliste** für die erste Iteration dieses inkrementellen Verfahrens ausgewählt. 
Im Anschluss an die [Analyse](#analyse-und-strukturierung) des Altsystems, wo die zu migrierenden Bestandteile und betroffenen *Java-Klassen* identifiziert wurden, erfolgte zunächst die [Entwicklung](#entwurf-und-entwicklung-der-neuen-benutzeroberflche) dieser Vertragsliste in Form einer **React App**. 
Anschließend wurde eine [Schnittstelle in Form einer **REST API**](#entwicklung-einer-rest-api-als-schnittstelle-zwischen-alt-und-neu) entwickelt, um die Kommunikation zwischen dem Altsystem und der migrierten Komponente herzustellen. 
Mit der [Anbindung der **React App** an die Schnittstelle](#anbindung-der-react-app-an-die-api) und dem Austausch der alten Komponente durch die neue *React App* wurde der erste Migrationsschritt [erfolgreich abgeschlossen](#migration-und-umstellung-des-frontends) . 

> Das in diesem Artikel demonstrierte Prozedere ist für jede weitere Komponente des Frontends zu wiederholen, bis dass das Frontend mit der letzten Komponente vollständig migriert wurde. 

Reviewer: ```suggestion
Das in diesem Artikel demonstrierte Prozedere ist für jede weitere Komponente des Frontends zu wiederholen, bis das Frontend mit der letzten Komponente vollständig migriert wurde. 
```

---
layout: [post, post-xml]             
title:  "Frontend Migration mittels der Chicken-Little-Methode"
date:   2020-03-16 01:23             
author: mschick                     
categories: [Methodik]                   
tags: [Frontend, Migration, Softwaremigration, React, ReactJS, JSX, JSF, Facelets, REST]  
---

> In meinem [ersten Beitrag](/_posts/2020-03-16-inkrementelle-softwaremigration-nach-der-chicken-little-methode.md) habe ich die **Chicken-Little-Methode**, eine Vorgehensweise für die schrittweise Migration von *Legacy-Systemen*, vorgestellt.

Reviewer 2: Hier ist der
```suggestion
In meinem [ersten Beitrag](https://www.adesso.de/de/news/blog/inkrementelle-softwaremigration-nach-der-chicken-little-methode-4.jsp) habe ich die **Chicken-Little-Methode**, eine Vorgehensweise für die schrittweise Migration von *Legacy-Systemen*, vorgestellt.
```

        </rich:column>
        ...
    </rich:dataTable>
    ...
    <h:commandButton id="berechnenButton"
        action="#{vertragslisteController.berechneVertraege()}"
        value="Berechnen" />

</ui:define>
```

#### Involvierte Java-Klassen
Aus dem oben stehenden *JSF-Template* lassen sich bereits die beteiligten *Java-Klassen* ablesen.
Die zentralen Elemente sind hierbei der Controller `VertragslisteController`, der das Datenhandling übernimmt, sowie die Entität `Vertrag`, die den jeweiligen Versicherungsvertrag repräsentiert und dessen Details bereitstellt.

## Entwurf und Entwicklung der neuen Benutzeroberfläche

Die Darstellung der *Vertragsliste* innerhalb der *React App* soll auch hier wieder in tabellarischer Form erfolgen.  

Sowohl die *Vertragsliste* als auch der *Vertrag* werden dazu jeweils als eigenständige *React Komponente* innerhalb der *React App* implementiert.
Bei einer *React Komponente* handelt es sich knapp zusammengefasst um einen wiederverwendbaren Teil einer Benutzeroberfläche, der die Funktionalität sowie die Eigenschaften der Komponente (`props`) innerhalb einer *React App* bündelt.
Ihr Aufbau ähnelt stark dem einer *(Java-)Klasse*.  

> Weitere Details dazu findet man u.a. [hier](https://reactjs.org/docs/components-and-props.html).

Reviewer 2: ```suggestion
Weitere Details dazu finden sich [in der React-Dokumantion](https://reactjs.org/docs/components-and-props.html).
```

                        {vertraege && vertraege.length > 0 ? (
                            <table>
                                <colgroup span="8"></colgroup>
                                <thead>{!-- Tabellenkopf -->}</thead>
                                <tbody>{vertraege.map(this.eachVertrag)}</tbody>
                            </table>
                        ) : (
                            <div>Keine Verträge gefunden</div>
                        )}

                        {vertraege && vertraege.length > 0 &&
                        <div className="actionButtons">
                            <input id="calculate"
                                   type="submit"
                                   name="calculate"
                                   value="Berechnen"
                                   onClick={this.calculate}/>
                        </div>
                        }
                    </div>
                )
            }
        }
}
```

**Randbemerkung:** Das, was im obigen Beispiel wie `HTML-Markup` anmutet ist in Wirklichkeit `JavaScript-Code` in Form von `JSX`.
Der `JSX-Code` wird durch einen *Transpiler* in *JavaScript* übersetzt.

> Entworfen wurde das ganze von Facebook als Erweiterung des **ES6-Standards**.  Weitere Informationen zu `JSX` und was dahinter steckt, findet Ihr [hier](https://reactjs.org/docs/jsx-in-depth.html) und [hier](https://facebook.github.io/jsx/).

Reviewer 2: Die Links bitte auch mit einem sprechenden Text versehen.

    }

    @POST
    public Response berechneAngebote(@QueryParam("cid") String cid) {
        try {
            vertragslisteController.berechneAngebote();
        } catch (Exception e) {
            log.error(e);
            return Response.serverError().build();
        }

        return Response.ok().build();
    }

    @DELETE
    @Path("/{vertragsId}")
    public Response removeVertrag(@PathParam("vertragsId") int vertragsId) {

        vertragslisteController.removeVertrag(vertragsId);
        return Response.noContent().build();

    }
}
```      

Für die Umsetzung wird sich der Library **RESTEasy** bedient, die den **JAX-RS**-Standard implementiert und die Definition und Konfiguration von *REST-Ressourcen* über Annotationen der *Java-Klassen* erlaubt. 

Über dem Klassennamen erfolgt zunächst die Definition der URI `/vertragsliste` per Annotation.
Darüber lässt sich die Ressource in der **REST-API** adressieren.
Daneben wird außerdem festgelegt, dass die Kommunikation via `JSON` erfolgen soll. 

Die folgende Tabelle listet die beteiligten **Ressourcen**, die jeweilige **URI**, die dabei verwendete **HTTP-Methode** sowie die jeweils korrespondierende **Java-Methode** auf.

> | Aktion                 | HTTP-Methode       | URI                           | Java-Methode           |

Reviewer 2: Die Tabellle ist noch in der Vorschau kaputt. @jo2 Schaust Du Dir das einmal an?

Reviewer: Bin leider erst jetzt dazu gekommen. @acetous Gibts nen Style-Guide oder ähnliches wie eine Tabelle aussehen soll? Konnte bisher keinen Blogeintrag mit tabelle finden?

Reviewer 2: Tabellen sind aktuell nicht möglich, daher muss das leider anders dargestellt werden.

Iframes passen sich standardmäßig nicht der Größe ihres Inhalts an. 
Das führt zu unschönen Effekten wie beispielsweise doppelten Scrollbars, einmal für die Seite selbst und dann nochmal für den Iframe-Inhalt. 

Um diesem Problem zu begegnen und den Einsatz von Iframes für den Benutzer zu kaschieren, gibt es verschiedene *JavaScript Libraries* (wie bspw. [Seamless.js](https://github.com/travist/seamless.js) oder [IframeResizer](https://github.com/davidjbradshaw/iframe-resizer)), die diese Dynamik für Iframes nachrüsten.   

In diesem Beispiel wurde die Librarie `IframeResizer` verwendet. 
Das unten stehende Quellcode-Beispiel zeigt die Integration:

```xml
<ui:define name="content">
    <iframe id="vertragsliste-react" src="https://makler.portal/api/vertragsliste"></iframe>
    
    <script>
        // IframeResizer
        iFrameResize({ inPageLinks: true }, "#vertragsliste-react");
    </script>
</ui:define>
```

Mit diesem letzten Schritt ist die Migration der Vertragsliste abgeschlossen und kann im Altsystem nun verwendet werden.

> # Zusammenfassung

Reviewer 2: Bitte die Links auf die einzelnen Absätze entfernen, das klappt im Blog später nicht.
